core/http/client/src/main/java/org/eclipse/rdf4j/http/client/BackgroundResultExecutor.java

```diff
@@ -85,4 +94,151 @@ private void autoCloseRunnable(QueryResult<?> result, Runnable runner) {
 			}
 		});
 	}
+
+	private static class CleanerTupleQueryResult implements TupleQueryResult {
+
+		private static final Logger logger = LoggerFactory.getLogger(CleanerTupleQueryResult.class);
+
+		private final TupleQueryResult delegate;
+		private final Cleaner.Cleanable cleanable;
+		private final CleanableState state;
+
+		public CleanerTupleQueryResult(TupleQueryResult delegate, ConcurrentCleaner cleaner) {
+			this.delegate = delegate;
+			this.state = new CleanableState(delegate);
+			this.cleanable = cleaner.register(this, state);
+		}
+
+		@Override
+		public void close() {
+			state.close();
+			cleanable.clean();
+		}
+
+		@Override
+		public boolean hasNext() {
+			return delegate.hasNext();
+		}
+
+		@Override
+		public BindingSet next() {
+			return delegate.next();
+		}
+
+		@Override
+		public void remove() {
+			delegate.remove();
+		}
+
+		@Override
+		public List<String> getBindingNames() throws QueryEvaluationException {
+			return delegate.getBindingNames();
+		}
+
+		private final static class CleanableState implements Runnable {
+
+			private final TupleQueryResult iteration;
+			private boolean closed = false;
```

Comment:
Should this not be volatile?

Reply:
It is only used from the `run()` method to reduce the chances that the cleaner closes the underlying iteration superfluously. If one thread closes the query result (which writes `true` to the `closed` field) and garbage collection runs  immediately afterwards and triggers the cleaner, then there is some chance that the write to the `closed` field will not yet be visible. I don't think that it's a scenario that is worth catering to by adding the overhead of a volatile write when closing the query results.